### PR TITLE
feat(buttonwithdialog): add focuslock

### PR DESCRIPTION
**Pull request by jamesb3ll**
This change is [Reviewable](https://reviewable.io/reviews/qantasairways/runway/162)

### CODE REVIEW

File: src/components/FocusLock/index.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import React, { Component, Children } from 'react';
```
Comment (rachmulvey) [inline src/components/FocusLock/index.js]: 😍

File: src/components/FocusLock/index.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import { findDOMNode } from 'react-dom';
+import PropTypes from 'prop-types';
+
+// From: https://github.com/davidtheclark/tabbable
+const tabbableSelector = [
```
Comment (tysonwolker) [inline src/components/FocusLock/index.js]: What about for hidden elements/tabIndex?

Comment (tysonwolker) [reply src/components/FocusLock/index.js]: Also input[type=hidden] needs to be excluded too.

Comment (tysonwolker) [reply src/components/FocusLock/index.js]: And non native elements ones that have tab-index added for example

**Comment by rachmulvey** [comment]
dismissed my approval because Tyson makes a good point

File: src/components/FocusLock/index.js (hunk 2 above)
Comment (jamesb3ll) [reply src/components/FocusLock/index.js]: Thanks for the help getting this right!

**Comment by jamesb3ll** [comment]
Thanks for the review, I sat with Tyson yesterday and worked out the edge cases for hidden elements (`type="hidden"`, `display: none` or `visibility: hidden`)

File: src/components/FocusLock/index.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    // Using findDOMNode as React doesn't support Fragment refs (yet?)
+    // https://github.com/facebook/react/pull/13841#issuecomment-430066195
+    // eslint-disable-next-line react/no-find-dom-node
+    this.base = findDOMNode(this);
+
+    const tabbables = [...this.base.querySelectorAll(tabbableSelector)];
+
+    // First tabbable element that is visible
+    this.firstFocusEl = tabbables.find(
```
Comment (tysonwolker) [inline src/components/FocusLock/index.js]: Perfect 👌

File: src/constants/keyCodes.js [hunk 4]
```diff
@@ -5,3 +5,4 @@ export const KEY_CODE_DOWN = 40;
 export const KEY_CODE_SPACE = 32;
 export const KEY_CODE_ENTER = 13;
 export const KEY_CODE_TAB = 9;
+export const KEY_CODE_ESC = 27;
```
Comment (tysonwolker) [inline src/constants/keyCodes.js]: Above and beyond 😉

File: src/components/Typeahead/index.js [hunk 5]
```diff
@@ -273,25 +262,14 @@ class Typeahead extends Component {
...
                     },
                     onBlur: () => {
-                      setState({
-                        inputValue: itemToString(selectedItem),
-                        placeholder: ''
-                      });
-                      this.setState({
-                        placeholder: ''
-                      });
+                      setState({ inputValue: itemToString(selectedItem) });
```
Comment (congcongcong250) [inline src/components/Typeahead/index.js]: thanks for clearing the logic, it took me some time to understand

**Comment by congcongcong250** [comment]
Can you close my focus outside pr when merging this one?

**Pull request merged**